lib/benchee/statistics/percentile.ex

```diff
@@ -1,50 +1,67 @@
 defmodule Benchee.Statistics.Percentile do
   @moduledoc false
 
+  @type percentile :: float()
+  @type percentiles :: %{optional(number()) => percentile()}
+
   @doc """
-  Calculates the value at the `percentile_number`-th percentile. Think of this as the
-  value below which `percentile_number` percent of the samples lie. For example,
+  Calculates the value at the `percentile_rank`-th percentile. Think of this as the
+  value below which `percentile_rank` percent of the samples lie. For example,
   if `Benchee.Statistics.Percentile.percentile(samples, 99)` == 123.45,
   99% of samples are less than 123.45.
 
-  ## Examples
-
-  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 8, 100)
-  5.0
-
-  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 8, 150)
-  5.0
+  Passing a number for `percentile_rank` calculates a single percentile.
+  Passing a list of numbers calculates multiple percentiles, and returns them
+  as a map like %{90 => 45.6, 99 => 78.9}, where the keys are the percentile
+  numbers, and the values are the percentile values.
 
-  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 8, 0)
-  1.0
+  ## Examples
 
-  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 8, -1)
+  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 12.5)
   1.0
 
   iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 50)
   3.0
 
   iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 75)
   4.75
+
+  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 99)
+  5.0
+
+  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], [50, 75, 99])
+  %{50 => 3.0, 75 => 4.75, 99 => 5.0}
+
+  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 100)
+  ** (ArgumentError) percentile must be between 0 and 100, got: 100
+
+  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 0)
+  ** (ArgumentError) percentile must be between 0 and 100, got: 0
   """
-  @spec percentile(list(number()), integer()) :: float()
-  def percentile(samples, percentile_number) do
-    percentile(samples, length(samples), percentile_number)
+  @spec percentile(list(number()), number() | list(number())) ::
```

Comment:
do we need the parentheses? Looks a bit... distracting.

Reply:
We're getting them with the formatter whether we like it or not, so I'd say we keep them. We could use the `[number()]` syntax for the lists to remove a couple parentheses.

Reply:
I only added those because I noticed in my elixir code formatter pull requests, the code formatter was doing that. Or I thought so. @devonestes you wrote an article—does the code formatter add parentheses to types? :)

Reply:
@devonestes thanks for answering my question......before I asked it!

Reply:
I'm already starting to _"like"_ the formatter... ;P

(seriously though, why do types in type specs have to look like function calls... makes little sense to me, maybe Jose has a rationale somewhere (I'm sure he has, question is just if he wrote it down somewhere))